go/libraries/doltcore/doltdb/local_table_editor.go

```diff
@@ -0,0 +1,309 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doltdb
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/liquidata-inc/dolt/go/libraries/doltcore/row"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+type LocalTableEditor struct {
+	restrictedDelete  bool
+	restrictedUpdate  bool
+	tableEditor       *TableEditor
+	tablesToReference []*ForeignKey
+	tablesToUpdate    []*ForeignKey
+}
+
+// ApplyToRoot is a shortcut to calling GlobalTableEditor.ApplyToRoot.
+func (lte *LocalTableEditor) ApplyToRoot(ctx context.Context, root *RootValue) (*RootValue, error) {
+	return globalTableEditor.ApplyToRoot(ctx, root)
+}
+
+// ContainsIndexedKey returns whether the given key is contained within the index. The key is assumed to be in the
+// format expected of the index, similar to searching on the index map itself.
+func (lte *LocalTableEditor) ContainsIndexedKey(ctx context.Context, key types.Tuple, indexName string) (bool, error) {
+	return lte.tableEditor.ContainsIndexedKey(ctx, key, indexName)
+}
+
+// GetIndexedRows returns all matching rows for the given key on the index. The key is assumed to be in the format
+// expected of the index, similar to searching on the index map itself.
+func (lte *LocalTableEditor) GetIndexedRows(ctx context.Context, key types.Tuple, indexName string) ([]row.Row, error) {
+	return lte.tableEditor.GetIndexedRows(ctx, key, indexName)
+}
+
+// GetRow returns the row matching the key given from the TableEditor. This is equivalent to calling Table and then
+// GetRow on the returned table, but a tad faster.
+func (lte *LocalTableEditor) GetRow(ctx context.Context, key types.Tuple) (row.Row, bool, error) {
+	return lte.tableEditor.GetRow(ctx, key)
+}
+
+// InsertRow adds the given row to the table. If the row already exists, use UpdateRow.
+func (lte *LocalTableEditor) InsertRow(ctx context.Context, dRow row.Row) error {
+	globalTableEditor.writeMutex.RLock()
+	defer globalTableEditor.writeMutex.RUnlock()
+
+	if !globalTableEditor.DisableForeignKeyChecks {
```

Comment:
Move this check inside validToInsert. Always call validate, let it decide which checks to perform.